src/tm-threads.c

```diff
@@ -2064,27 +2064,46 @@ static void TmThreadDumpThreads(void)
 }
 #endif
 
+/* Aligned to CLS to avoid false sharing between atomic ops. */
 typedef struct Thread_ {
     ThreadVars *tv;     /**< threadvars structure */
     const char *name;
     int type;
     int in_use;         /**< bool to indicate this is in use */
 
-    SCTime_t pktts;         /**< current packet time of this thread
-                             *   (offline mode) */
+    SC_ATOMIC_DECLARE(SCTime_t, pktts); /**< current packet time of this thread
+                                         *   (offline mode) */
     SCTime_t sys_sec_stamp; /**< timestamp in real system
                              *   time when the pktts was last updated. */
-} Thread;
+    SCSpinlock spin;
+} __attribute__((aligned(CLS))) Thread;
 
 typedef struct Threads_ {
     Thread *threads;
     size_t threads_size;
     int threads_cnt;
 } Threads;
 
+static bool thread_store_sealed = false;
 static Threads thread_store = { NULL, 0, 0 };
 static SCMutex thread_store_lock = SCMUTEX_INITIALIZER;
 
+void TmThreadsSealThreads(void)
+{
+    SCMutexLock(&thread_store_lock);
+    DEBUG_VALIDATE_BUG_ON(thread_store_sealed);
+    thread_store_sealed = true;
+    SCMutexUnlock(&thread_store_lock);
+}
+
```

Comment:
Sealing appears to be only checked under DEBUG_VALIDATE.  Is this meany a development debugging tool to make sure thread fields are accessed by multiple threads before they are ready to be?

Reply:
Not entirely sure what you're asking, but the goal is to index the array safely after all threads have been set up. Since the array is dynamically allocated and grown (realloc) as threads are spawning, we need to know for sure when it is ready. The debug validation is there to make sure we can catch it if the order is wrong, so we access w/o lock while unsealed.

Reply:
It appears the "sealing" only sets a flag, and nothing else. And that flag is only checked with DEBUG_VALIDATE_BUG_ON. Is there more to it?

Reply:
no, the debug validation check is there to make sure it is respected

Reply:
Ok, as threads are only registered at startup, why not always check it with a fatal error instead of having to opt-in with debug-validation.

Reply:
Still just on setup and teardown right?  I guess my line of thinking is if its important enough to seal, its important enough to check - always.